core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java

```diff
@@ -0,0 +1,213 @@
+      }
+
+      // Check that the field is of primitive type
+      Schema fieldSchema = inputSchemaField.getSchema();
+      Schema.Type fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+      if (!fieldType.isSimpleType()) {
+        throw new IllegalArgumentException(
+          "Type for field  " + config.bodyField + " must be primitive type");
+      }
+    }
+  }
+
+  @Override
+  public void initialize(RealtimeContext context) throws Exception {
+    super.initialize(context);
+    BasicAWSCredentials awsCred = new BasicAWSCredentials(config.awsAccessKey, config.awsAccessSecret);
+    kinesisClient = new AmazonKinesisClient(awsCred);
+    createAndWaitForStreamToBecomeAvailable(kinesisClient,
+                                            config.name,
+                                            config.shardCount);
+  }
+
+  @Override
+  public int write(Iterable<StructuredRecord> structuredRecords, DataWriter dataWriter) throws Exception {
+    int numRecordsWritten = 0;
+    for (StructuredRecord structuredRecord : structuredRecords) {
+      Schema schema = structuredRecord.getSchema();
+      Object data = structuredRecord.get(config.bodyField);
+      if (data == null) {
+        LOG.debug("Found null data. Skipping record.");
+        continue;
+      }
+      Schema.Field dataSchemaField = schema.getField(config.bodyField);
+      PutRecordRequest putRecordRequest = new PutRecordRequest();
+      putRecordRequest.setStreamName(config.name);
+      putRecordRequest.setPartitionKey(config.partition);
+      switch (dataSchemaField.getSchema().getType()) {
+        case BYTES:
+          numRecordsWritten += writeBytes(putRecordRequest, data);
+          break;
+        case STRING:
+          numRecordsWritten += writeString(putRecordRequest, data);
+          break;
+        default:
+          if (dataSchemaField.getSchema().getType().isSimpleType()) {
+            numRecordsWritten += writeBytes(putRecordRequest, String.valueOf(data));
+            break;
+          }
+          LOG.debug("Type {} is not supported for writing to stream", data.getClass().getName());
+          break;
+      }
+    }
+    return numRecordsWritten;
+  }
+
+  private int writeString(PutRecordRequest putRecordRequest, Object data) {
+    putRecordRequest.setData(ByteBuffer.wrap(Bytes.toBytes((String) data)));
+    try {
+      PutRecordResult result = kinesisClient.putRecord(putRecordRequest);
+      LOG.info("Data written at {} ", result);
+      return 1;
+    } catch (Exception e) {
+      LOG.warn("could not write data to stream {}", putRecordRequest.getStreamName(), e);
+      return  0;
+    }
+  }
+
+  private int writeBytes(PutRecordRequest putRecordRequest, Object data) {
+    ByteBuffer buffer;
+    if (data instanceof ByteBuffer) {
+      buffer = (ByteBuffer) data;
+    } else if (data instanceof byte[]) {
+      buffer = ByteBuffer.wrap((byte[]) data);
+    } else {
```

Comment:
this should never happen right? this method is always passed a byte array or a byte buffer? Just throw an IllegalStateException from the else, otherwise it seems like this is a valid scenario while reading the code.